cpp/src/neighbors/detail/vamana/vamana_build.cuh

```diff
@@ -0,0 +1,471 @@
+			raft::make_host_scalar_view(&n_iter));
+			
+
+  auto h_centroid = raft::make_host_matrix<T,int64_t>(1, dim);
+  raft::copy(h_centroid.data_handle(), centroids.data_handle(), centroids.size(), stream);
+
+  printf("centroid\n");
+  for(int i=0; i<dim; i++){
+    printf("%f, ", h_centroid(0,i));
+  }
+  printf("\n");
+
+//  rmm::device_uvector<DistPair<IdxT,accT>> d_medoid(
+//        num_reduc_threads, stream, raft::resource::get_large_workspace_resource(res));
+
+
+  IdxT final_medoid=0;
+  return final_medoid;
+ 
+}
+*/
+
+/********************************************************************************************
+ * Main Vamana building function - insert vectors into empty graph in batches
+ * Pre - dataset contains the vector data, host matrix allocated to store the graph
+ * Post - graph matrix contains the graph edges of the final Vamana graph
+ *******************************************************************************************/
+template<typename T,
+         typename accT,
+         typename IdxT = uint32_t,
+         typename Accessor = raft::host_device_accessor<std::experimental::default_accessor<T>,
+                                                         raft::memory_type::host>>
+void batched_insert_vamana(
+  raft::resources const& res,
+  const index_params& params,
+  raft::mdspan<const T, raft::matrix_extent<int64_t>, raft::row_major, Accessor> dataset,
+  raft::host_matrix_view<IdxT, int64_t> graph,
+  cuvs::distance::DistanceType metric, int dim)
+{
+  auto stream = raft::resource::get_cuda_stream(res);
+  int N = dataset.extent(0);
+  int degree = graph.extent(1);
+
+  // Algorithm params
+  int max_batchsize = (int)(params.max_batchsize*(float)N);
+  if (max_batchsize > (int)dataset.extent(0)) {
+    RAFT_LOG_WARN(
+      "Max batch insert size cannot be larger 1.0, reducing it to 1");
+    max_batchsize = (int)dataset.extent(0);
+  }
+  int insert_iters = (int)(params.vamana_iters);
+  double base = (double)(params.batch_base);
+  float alpha = (float)(params.alpha);
+  int visited_size = params.visited_size;
+  int queue_size = params.queue_size;
+
+  // create gpu graph and set to all -1s
+  auto d_graph = raft::make_device_matrix<IdxT, int64_t>(res, graph.extent(0), graph.extent(1));
+  memset_graph<IdxT><<<256,blockD>>>(d_graph.view());
+
+  // Temp storage about each batch of inserts being performed
+  auto query_ids = raft::make_device_vector<IdxT>(res, max_batchsize);
+  rmm::device_buffer query_list_ptr{
+      (max_batchsize+1)*sizeof(QueryCandidates<IdxT, accT>), stream, raft::resource::get_large_workspace_resource(res)};
+  QueryCandidates<IdxT,accT>* query_list = static_cast<QueryCandidates<IdxT,accT>*>(query_list_ptr.data());
+
+  
+  // Results of each batch of inserts during build - Memory is used by query_list structure
+  rmm::device_uvector<IdxT> visited_ids(
+       max_batchsize*visited_size, stream, raft::resource::get_large_workspace_resource(res));
```

Comment:
Moved everything to use raft mdarray structures. I did not realize that the large_workspace is frequently managed memory... This may have a big impact on perf, and I'm not 100% sure which structures to make sure are not managed and which to allow this for. May need to do some testing on perf impact of this.